File: csharp/cars-assemble/CarsAssemble.cs
using System;

static class AssemblyLine
{
	const int CarsPerHour = 221;
	public static double SuccessRate(int speed)
	{
		if (speed == 10)
		{
			return 0.77;
		}
		else if (speed == 9)
		{
			return 0.8;
		}
		else if (speed >= 5 && speed <= 8)
		{
			return 0.9;
		}
		else if (speed >= 1 && speed <= 4)
		{
			return 1;
		}
		else
		{
			return 0;
		}
	}

	public static double ProductionRatePerHour(int speed)
	{
		return SuccessRate(speed) * CarsPerHour * speed;
	}

	public static int WorkingItemsPerMinute(int speed)
	{
		return (int)(ProductionRatePerHour(speed) / 60);
	}
}

File: go/cars-assemble/cars_assemble.go
package cars

const minsPerHour = 60

// carsPerHour is amount of cars produced per hour for speed = 1
const carsPerHour = 221

// SuccessRate is used to calculate the ratio of an item being created without
// error for a given speed
func SuccessRate(speed int) float64 {
	if speed == 0 {
		return 0.0
	} else if 1 <= speed && speed <= 4 {
		return 1.0
	} else if speed >= 5 && speed <= 8 {
		return 0.9
	} else {
		return 0.77
	}
}

// CalculateProductionRatePerHour for the assembly line, taking into account
// its success rate
func CalculateProductionRatePerHour(speed int) float64 {
	return float64(speed) * SuccessRate(speed) * carsPerHour
}

// CalculateProductionRatePerMinute describes how many working items are
// produced by the assembly line every minute
func CalculateProductionRatePerMinute(speed int) int {
	return int(CalculateProductionRatePerHour(speed) / minsPerHour)
}

// CalculateLimitedProductionRatePerHour describes how many working items are
// produced per hour with an upper limit on how many can be produced per hour
func CalculateLimitedProductionRatePerHour(speed int, limit float64) float64 {
	rate := CalculateProductionRatePerHour(speed)
	if rate > limit {
		rate = limit
	}
	return rate
}
